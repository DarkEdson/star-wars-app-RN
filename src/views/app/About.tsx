import * as React from 'react';
import {Button} from 'react-native';
import {useNavigation} from '@react-navigation/native';
import {HomeScreenNavigationProp} from '../../navigation/App';
import Container from '../../components/Container';
import Text from '../../components/Text';

export default function SettingScreen() {
  const navigate = useNavigation<HomeScreenNavigationProp>();
  return (
    <Container>
      <Text>Contacto: itic.rodrigo@gmail.com</Text>
      <Button
        title="ConfiguraciÃ³n"
        onPress={() => navigate.navigate('Settings', {id: '123'})}
      />
    </Container>
  );
}
