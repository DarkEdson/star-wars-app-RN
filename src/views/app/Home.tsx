import * as React from 'react';
import dayjs from 'dayjs';
import relativeTime from 'dayjs/plugin/relativeTime';
import Text from '../../components/Text';
import 'dayjs/locale/es-mx';
import Container from '../../components/Container';
import {Button} from 'react-native';
import {setSignOut} from '../../redux/slices/auth.slice';
import {useDispatch} from 'react-redux';
dayjs.extend(relativeTime);
dayjs.locale('es-mx');

export default function HomeScreen() {
  const dispatch = useDispatch();
  const doLogout = () => {
    console.log('test');
    dispatch(setSignOut());
  };
  return (
    <Container>
      <Text>
        Faltan {dayjs(`${new Date().getFullYear()}-12-25`).fromNow(true)} para
        navidad
      </Text>
      <Button title="Cerrar sesiÃ³n" onPress={() => doLogout()} />
    </Container>
  );
}
