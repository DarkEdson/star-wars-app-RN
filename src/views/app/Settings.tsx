import * as React from 'react';
import {StyleSheet, View} from 'react-native';
import {useNavigation} from '@react-navigation/native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import {useDispatch, useSelector} from 'react-redux';
import {HomeScreenNavigationProp} from '../../navigation/App';
import Container from '../../components/Container';
import Text from '../../components/Text';
import Button from '../../components/Button';
import {selectTheme, setTheme} from '../../redux/slices/app.slice';
import {colors} from '../../theme';

export default function SettingScreen() {
  const navigate = useNavigation<HomeScreenNavigationProp>();
  const currentTheme = useSelector(selectTheme);
  const dispatch = useDispatch();
  const changeTheme = () => {
    dispatch(setTheme(currentTheme === 'light' ? 'dark' : 'light'));
  };

  const styles = StyleSheet.create({
    container: {
      justifyContent: 'center',
      alignItems: 'center',
      flexDirection: 'row',
      gap: 10,
    },
  });
  return (
    <Container>
      <Text>Pantalla de configuración</Text>
      <Button testID="goBackButton" onPress={() => navigate.goBack()}>
        <Text buttonText>Atrás</Text>
      </Button>
      <Button testID="themeButton" onPress={changeTheme}>
        {currentTheme === 'light' ? (
          <View style={styles.container}>
            <Text testID="darkText" buttonText>
              Activar modo oscuro
            </Text>
            <Icon
              name="dark-mode"
              size={20}
              color={colors[currentTheme].buttonText}
            />
          </View>
        ) : (
          <View style={styles.container}>
            <Text testID="lightText" buttonText>
              Activar modo claro
            </Text>
            <Icon
              name="light-mode"
              size={20}
              color={colors[currentTheme].buttonText}
            />
          </View>
        )}
      </Button>
    </Container>
  );
}
