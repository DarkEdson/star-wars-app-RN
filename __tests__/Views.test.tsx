import {it, afterEach, expect} from '@jest/globals';
import 'react-native';
import React from 'react';
import {cleanup, fireEvent, render, waitFor} from './utils/test.utils';
import HomeScreen from '../src/views/app/Home';
import AboutScreen from '../src/views/app/About';
import SettingScreen from '../src/views/app/Settings';
import LoginScreen from '../src/views/auth/Login';
import {NavigationContainer} from '@react-navigation/native';
import {setTheme} from '../src/redux/slices/app.slice';

const mockDispatch = jest.fn();

const mockedNavigate = jest.fn();

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}));

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
      goBack: mockedNavigate,
    }),
  };
});

afterEach(() => {
  cleanup();
});

describe('Renders Home Screen correctly', () => {
  it('Renders Home Screen correctly', async () => {
    let component;
    await waitFor(() => {
      component = render(<HomeScreen />, {});
    });

    expect(component!.toJSON()).toMatchSnapshot();
  });

  it('Click logout button functionality', async () => {
    let component: any;
    await waitFor(() => {
      component = render(
        <NavigationContainer>
          <HomeScreen />
        </NavigationContainer>,
        {},
      );
    });
    fireEvent.press(await component.findByTestId('logoutButton'));
    expect(mockDispatch).toBeCalled();
  });
});

describe('Renders About Screen correctly', () => {
  it('Renders About Screen correctly', async () => {
    let component;
    await waitFor(() => {
      component = render(
        <NavigationContainer>
          <AboutScreen />
        </NavigationContainer>,
        {},
      );
    });

    expect(component!.toJSON()).toMatchSnapshot();
  });

  it('Call button to navigate', async () => {
    let component: any;
    await waitFor(() => {
      component = render(
        <NavigationContainer>
          <AboutScreen />
        </NavigationContainer>,
        {},
      );
    });
    fireEvent.press(component.getByTestId('goToSettingsButton'));
    expect(mockedNavigate).toBeCalled();
  });
});

describe('Renders Settings Screen correctly', () => {
  it('Renders Settings Screen correctly', async () => {
    let component;
    await waitFor(() => {
      component = render(
        <NavigationContainer>
          <SettingScreen />
        </NavigationContainer>,
        {},
      );
    });

    expect(component!.toJSON()).toMatchSnapshot();
  });

  it('Check light theme exists', async () => {
    let component: any;
    await waitFor(() => {
      component = render(
        <NavigationContainer>
          <SettingScreen />
        </NavigationContainer>,
        {
          actionToDispatch: setTheme('dark'),
        },
      );
    });
    const darkButton = await component.findByTestId('lightText');
    expect(darkButton).toBeTruthy();
  });

  it('Check dark theme exists', async () => {
    let component: any;
    await waitFor(() => {
      component = render(
        <NavigationContainer>
          <SettingScreen />
        </NavigationContainer>,
        {
          actionToDispatch: setTheme('light'),
        },
      );
    });
    const darkButton = await component.findByTestId('darkText');
    expect(darkButton).toBeTruthy();
  });

  it('click button to change theme to dark', async () => {
    let component: any;
    await waitFor(() => {
      component = render(
        <NavigationContainer>
          <SettingScreen />
        </NavigationContainer>,
        {
          actionToDispatch: setTheme('light'),
        },
      );
    });
    fireEvent.press(await component.findByTestId('themeButton'));
    expect(mockDispatch).toBeCalled();
  });

  it('click button to change theme to light', async () => {
    let component: any;
    await waitFor(() => {
      component = render(
        <NavigationContainer>
          <SettingScreen />
        </NavigationContainer>,
        {
          actionToDispatch: setTheme('dark'),
        },
      );
    });
    fireEvent.press(await component.findByTestId('themeButton'));
    expect(mockDispatch).toBeCalled();
  });

  it('Go back', async () => {
    let component: any;
    await waitFor(() => {
      component = render(
        <NavigationContainer>
          <SettingScreen />
        </NavigationContainer>,
        {},
      );
    });
    fireEvent.press(component.getByText('AtrÃ¡s'));
    expect(mockedNavigate).toBeCalled();
  });
});

describe('Login screen', () => {
  it('Renders Login Screen correctly', async () => {
    let component;
    await waitFor(() => {
      component = render(
        <NavigationContainer>
          <LoginScreen />
        </NavigationContainer>,
        {},
      );
    });

    expect(component!.toJSON()).toMatchSnapshot();
  });

  it('Login on click button', async () => {
    let component: any;
    await waitFor(() => {
      component = render(
        <NavigationContainer>
          <LoginScreen />
        </NavigationContainer>,
        {},
      );
    });
    expect(component.getByTestId('loginButton')).toBeTruthy();
    fireEvent.press(component.getByText('Presiona para entrar'));
    expect(mockDispatch).toBeCalled();
  });
});
