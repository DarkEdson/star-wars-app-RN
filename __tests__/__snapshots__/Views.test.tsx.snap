// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login screen Renders Login Screen correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#242424",
      "height": 1334,
      "paddingLeft": 37.5,
      "paddingRight": 37.5,
      "paddingTop": 80,
      "width": 750,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#9D59FB",
        "justifyContent": "center",
        "marginVertical": 15,
        "opacity": 1,
        "paddingVertical": 10,
        "width": 675,
      }
    }
    testID="loginButton"
  >
    <Text
      style={
        {
          "color": "#FFF",
          "fontFamily": "Montserrat-Bold",
          "fontSize": 18,
        }
      }
    >
      <Text
        style={
          {
            "color": "#FFF",
            "fontFamily": "Montserrat-Regular",
            "fontSize": 18,
          }
        }
      >
        Presiona para entrar
      </Text>
    </Text>
  </View>
</View>
`;

exports[`Renders About Screen correctly Renders About Screen correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFF",
      "height": 1334,
      "paddingLeft": 37.5,
      "paddingRight": 37.5,
      "paddingTop": 80,
      "width": 750,
    }
  }
>
  <Text
    style={
      {
        "color": "#242424",
        "fontFamily": "Montserrat-Regular",
        "fontSize": 18,
      }
    }
  >
    Contacto: itic.rodrigo@gmail.com
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#9D59FB",
        "justifyContent": "center",
        "marginVertical": 15,
        "opacity": 1,
        "paddingVertical": 10,
        "width": 675,
      }
    }
    testID="goToSettingsButton"
  >
    <Text
      style={
        {
          "color": "#FFF",
          "fontFamily": "Montserrat-Bold",
          "fontSize": 18,
        }
      }
    >
      <Text
        style={
          {
            "color": "#FFF",
            "fontFamily": "Montserrat-Regular",
            "fontSize": 18,
          }
        }
      >
        Configuración
      </Text>
    </Text>
  </View>
</View>
`;

exports[`Renders Home Screen correctly Renders Home Screen correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFF",
      "height": 1334,
      "paddingLeft": 37.5,
      "paddingRight": 37.5,
      "paddingTop": 80,
      "width": 750,
    }
  }
>
  <Text
    style={
      {
        "color": "#242424",
        "fontFamily": "Montserrat-Regular",
        "fontSize": 18,
      }
    }
  >
    Faltan 
    10 months
     para navidad
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#9D59FB",
        "justifyContent": "center",
        "marginVertical": 15,
        "opacity": 1,
        "paddingVertical": 10,
        "width": 675,
      }
    }
    testID="logoutButton"
  >
    <Text
      style={
        {
          "color": "#FFF",
          "fontFamily": "Montserrat-Bold",
          "fontSize": 18,
        }
      }
    >
      <Text
        style={
          {
            "color": "#FFF",
            "fontFamily": "Montserrat-Regular",
            "fontSize": 18,
          }
        }
      >
        Cerrar sesión
      </Text>
    </Text>
  </View>
</View>
`;

exports[`Renders Settings Screen correctly Renders Settings Screen correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFF",
      "height": 1334,
      "paddingLeft": 37.5,
      "paddingRight": 37.5,
      "paddingTop": 80,
      "width": 750,
    }
  }
>
  <Text
    style={
      {
        "color": "#242424",
        "fontFamily": "Montserrat-Regular",
        "fontSize": 18,
      }
    }
  >
    Pantalla de configuración
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#9D59FB",
        "justifyContent": "center",
        "marginVertical": 15,
        "opacity": 1,
        "paddingVertical": 10,
        "width": 675,
      }
    }
    testID="goBackButton"
  >
    <Text
      style={
        {
          "color": "#FFF",
          "fontFamily": "Montserrat-Bold",
          "fontSize": 18,
        }
      }
    >
      <Text
        style={
          {
            "color": "#FFF",
            "fontFamily": "Montserrat-Regular",
            "fontSize": 18,
          }
        }
      >
        Atrás
      </Text>
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#9D59FB",
        "justifyContent": "center",
        "marginVertical": 15,
        "opacity": 1,
        "paddingVertical": 10,
        "width": 675,
      }
    }
    testID="themeButton"
  >
    <Text
      style={
        {
          "color": "#FFF",
          "fontFamily": "Montserrat-Bold",
          "fontSize": 18,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 10,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            {
              "color": "#FFF",
              "fontFamily": "Montserrat-Regular",
              "fontSize": 18,
            }
          }
          testID="darkText"
        >
          Activar modo oscuro
        </Text>
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "#FFF",
                "fontSize": 20,
              },
              undefined,
              {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          
        </Text>
      </View>
    </Text>
  </View>
</View>
`;
